File: python/2020/day_06.py
from pathlib import Path


def main():
    text = (Path(__file__).parent / "../input/input_06.txt").read_text()
    groups = text.split('\n\n')
    print("Part 1:", sum(len(set(g) - {'\n'}) for g in groups))
    print("Part 2:", sum(len(set.intersection(*(set(p) for p in g.splitlines()))) for g in groups))


if __name__ == "__main__":
    main()

File: rust/2020/src/bin/day_06.rs
use std::collections::HashSet;
use std::fs;

fn main() {
    let alpha: HashSet<_> = "abcdefghijklmnopqrstuvwxyz".chars().collect();
    let input = fs::read_to_string("../input/input_06.txt").unwrap();
    println!("Part 1: {}", input.split("\n\n").map(|g| {
        g.chars().filter(|&c| c != '\n').collect::<HashSet<_>>().len()
    }).sum::<usize>());
    println!("Part 2: {}", input.split("\n\n").map(|g| {
        g.lines().map(|p| p.chars().collect())
            .fold(alpha.clone(), |a, b| a.intersection(&b).cloned().collect())
            .len()
    }).sum::<usize>());
}
